contracts/integrations/test/utils/function_assertions.ts

```diff
@@ -0,0 +1,64 @@
+import { PromiseWithTransactionHash } from '@0x/base-contract';
+import { TransactionReceiptWithDecodedLogs } from 'ethereum-types';
+
+export interface ContractGetterFunction {
+    callAsync: (...args: any[]) => Promise<any>;
+    getABIEncodedTransactionData: (...args: any[]) => string;
+}
+
+export interface ContractWrapperFunction extends ContractGetterFunction {
+    awaitTransactionSuccessAsync?: (...args: any[]) => PromiseWithTransactionHash<TransactionReceiptWithDecodedLogs>;
+}
+
+export interface Result {
+    data?: any;
+    success: boolean;
+    receipt?: TransactionReceiptWithDecodedLogs;
+}
+
+export interface Condition {
+    before: (...args: any[]) => Promise<any>;
+    after: (beforeInfo: any, result: Result, ...args: any[]) => Promise<void>;
+}
+
+export class FunctionAssertion {
+    // A before and an after assertion that will be called around the wrapper function.
+    public condition: Condition;
+
+    // The wrapper function that will be wrapped in assertions.
+    public wrapperFunction: ContractWrapperFunction;
+
+    constructor(wrapperFunction: ContractWrapperFunction, condition: Condition) {
+        this.condition = condition;
+        this.wrapperFunction = wrapperFunction;
+    }
+
+    /**
+     * Runs the wrapped function and fails if the before or after assertions fail.
+     * @param ...args The args to the contract wrapper function.
+     */
+    public async runAsync(...args: any[]): Promise<void> {
```

Comment:
optional: As mentioned offline, it could enable some cool patterns to have `runAsync` return `beforeInfo` and `afterInfo`, should we change the return type of `after` to `Promise<any>`.

Reply:
I'm up for this change. This said, after using the framework some, I haven't had a need for this. I think we should re-evaluate whether being able to access these values is actually a good thing once we've used the framework more (we can access these values in `after` after all).